# Crude Oil Production Analysis

## Introduction

Crude oil production is a critical component of the global energy market and has significant implications for economies and industries worldwide. This project aims to analyze crude oil production data to uncover trends, patterns, and insights that can inform decision-making in the energy sector.

### Objectives
- **Data Collection**: Gather historical crude oil production data from reliable sources.
- **Data Cleaning**: Process the data to handle missing values, outliers, and inconsistencies.
- **Exploratory Data Analysis (EDA)**: Use statistical methods and visualizations to explore the data.
- **Trend Analysis**: Identify and analyze long-term trends in crude oil production.
- **Predictive Modeling**: Build models to forecast future production levels.

### Dataset
The dataset used in this project includes:
- Historical crude oil production data of Volve field.

### Tools and Technologies
- **Python**: Programming language used for data analysis and modeling.
- **Pandas**: Library for data manipulation and analysis.
- **Plotly.JS**: Libraries for data visualization.
- **Scikit-learn**: Machine learning library for predictive modeling.

### Structure of the Notebook
1. **Data Collection and Cleaning**: Steps to gather and preprocess the data.
2. **Exploratory Data Analysis**: Visualizations and statistical analysis of the data.
3. **Trend Analysis**: Examination of production trends over time.
4. **Predictive Modeling**: Development and evaluation of predictive models.
5. **Conclusions and Insights**: Key findings and their implications for the industry.

By the end of this project, we aim to provide a comprehensive analysis of crude oil production trends and deliver actionable insights that can help stakeholders in making informed decisions.


## Data Dictionary

| Column                  | Description                                                     |
|-------------------------|-----------------------------------------------------------------|
| DATEPRD                 | The date of the production data entry                           |
| WELL_BORE_CODE          | A unique identifier for the well bore                           |
| NPD_WELL_BORE_CODE      | The Norwegian Petroleum Directorate (NPD) specific code for the well bore |
| NPD_WELL_BORE_NAME      | The name of the well bore as per the NPD                        |
| NPD_FIELD_CODE          | The NPD code for the oil field                                  |
| NPD_FIELD_NAME          | The name of the oil field as per the NPD                        |
| NPD_FACILITY_CODE       | The NPD code for the facility                                   |
| NPD_FACILITY_NAME       | The name of the facility as per the NPD                         |
| ON_STREAM_HRS           | The number of hours the well was producing                      |
| AVG_DOWNHOLE_PRESSURE   | The average pressure measured downhole (in the wellbore)        |
| AVG_DOWNHOLE_TEMPERATURE| The average temperature measured downhole                       |
| AVG_DP_TUBING           | The average differential pressure in the tubing                 |
| AVG_ANNULUS_PRESS       | The average pressure in the annulus (the space between the wellbore and the tubing) |
| AVG_CHOKE_SIZE_P        | The average size of the production choke (controls flow rate)   |
| AVG_CHOKE_UOM           | The unit of measurement for the choke size                      |
| AVG_WHP_P               | The average wellhead pressure (pressure at the surface of the well) |
| AVG_WHT_P               | The average wellhead temperature                                |
| DP_CHOKE_SIZE           | The differential pressure across the choke                      |
| BORE_OIL_VOL            | The volume of oil produced                                      |
| BORE_GAS_VOL            | The volume of gas produced                                      |
| BORE_WAT_VOL            | The volume of water produced                                    |
| BORE_WI_VOL             | The volume of water injected                                    |
| FLOW_KIND               | The type of flow (e.g., production, injection)                  |
| WELL_TYPE               | The type of well (e.g., oil, gas, water injection)              |


## Importing libraries and getting started

In [34]:
import pandas as pd
import numpy as np

import plotly.express as px
import seaborn as sns

from sklearn.impute import MissingIndicator, KNNImputer
from sklearn.compose import make_column_selector, make_column_transformer
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.model_selection import train_test_split

seed = 0
np.random.seed(0)

pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', 500)

In [2]:
data = pd.read_excel('data/volve-field-daily-data.xlsx')

In [3]:
data.head()

DATEPRD WELL_BORE_CODE  NPD_WELL_BORE_CODE NPD_WELL_BORE_NAME  \
0 2014-04-07  NO 15/9-F-1 C                7405         15/9-F-1 C   
1 2014-04-08  NO 15/9-F-1 C                7405         15/9-F-1 C   
2 2014-04-09  NO 15/9-F-1 C                7405         15/9-F-1 C   
3 2014-04-10  NO 15/9-F-1 C                7405         15/9-F-1 C   
4 2014-04-11  NO 15/9-F-1 C                7405         15/9-F-1 C   

   NPD_FIELD_CODE NPD_FIELD_NAME  NPD_FACILITY_CODE NPD_FACILITY_NAME  \
0         3420717          VOLVE             369304    MÆRSK INSPIRER   
1         3420717          VOLVE             369304    MÆRSK INSPIRER   
2         3420717          VOLVE             369304    MÆRSK INSPIRER   
3         3420717          VOLVE             369304    MÆRSK INSPIRER   
4         3420717          VOLVE             369304    MÆRSK INSPIRER   

   ON_STREAM_HRS  AVG_DOWNHOLE_PRESSURE  AVG_DOWNHOLE_TEMPERATURE  \
0            0.0                0.00000                   0.00000   
1            0.0                    NaN                       NaN   
2            0.0                    NaN                       NaN   
3            0.0                    NaN                       NaN   
4            0.0              310.37614                  96.87589   

   AVG_DP_TUBING  AVG_ANNULUS_PRESS  AVG_CHOKE_SIZE_P AVG_CHOKE_UOM  \
0        0.00000                0.0          0.000000             %   
1            NaN                0.0          1.003059             %   
2            NaN                0.0          0.979008             %   
3            NaN                0.0          0.545759             %   
4      277.27826                0.0          1.215987             %   

   AVG_WHP_P  AVG_WHT_P  DP_CHOKE_SIZE  BORE_OIL_VOL  BORE_GAS_VOL  \
0    0.00000    0.00000        0.00000           0.0           0.0   
1    0.00000    0.00000        0.00000           0.0           0.0   
2    0.00000    0.00000        0.00000           0.0           0.0   
3    0.00000    0.00000        0.00000           0.0           0.0   
4   33.09788   10.47992       33.07195           0.0           0.0   

   BORE_WAT_VOL  BORE_WI_VOL   FLOW_KIND WELL_TYPE  
0           0.0          NaN  production        WI  
1           0.0          NaN  production        OP  
2           0.0          NaN  production        OP  
3           0.0          NaN  production        OP  
4           0.0          NaN  production        OP

In [4]:
data.nunique()

DATEPRD                     3327
WELL_BORE_CODE                 7
NPD_WELL_BORE_CODE             7
NPD_WELL_BORE_NAME             7
NPD_FIELD_CODE                 1
NPD_FIELD_NAME                 1
NPD_FACILITY_CODE              1
NPD_FACILITY_NAME              1
ON_STREAM_HRS                925
AVG_DOWNHOLE_PRESSURE       6567
AVG_DOWNHOLE_TEMPERATURE    6461
AVG_DP_TUBING               8684
AVG_ANNULUS_PRESS           6644
AVG_CHOKE_SIZE_P            6419
AVG_CHOKE_UOM                  1
AVG_WHP_P                   8829
AVG_WHT_P                   8793
DP_CHOKE_SIZE               9057
BORE_OIL_VOL                7818
BORE_GAS_VOL                8005
BORE_WAT_VOL                7361
BORE_WI_VOL                 5258
FLOW_KIND                      2
WELL_TYPE                      2
dtype: int64

In [5]:
data.shape

(15634, 24)

In [6]:
data.isna().sum()

DATEPRD                        0
WELL_BORE_CODE                 0
NPD_WELL_BORE_CODE             0
NPD_WELL_BORE_NAME             0
NPD_FIELD_CODE                 0
NPD_FIELD_NAME                 0
NPD_FACILITY_CODE              0
NPD_FACILITY_NAME              0
ON_STREAM_HRS                285
AVG_DOWNHOLE_PRESSURE       6654
AVG_DOWNHOLE_TEMPERATURE    6654
AVG_DP_TUBING               6654
AVG_ANNULUS_PRESS           7744
AVG_CHOKE_SIZE_P            6715
AVG_CHOKE_UOM               6473
AVG_WHP_P                   6479
AVG_WHT_P                   6488
DP_CHOKE_SIZE                294
BORE_OIL_VOL                6473
BORE_GAS_VOL                6473
BORE_WAT_VOL                6473
BORE_WI_VOL                 9928
FLOW_KIND                      0
WELL_TYPE                      0
dtype: int64

In [7]:
data.isna().sum().sum() / (data.shape[0]*data.shape[1]) * 100

22.33033772547013

In [8]:
data.isna().mean().round(4) * 100

DATEPRD                      0.00
WELL_BORE_CODE               0.00
NPD_WELL_BORE_CODE           0.00
NPD_WELL_BORE_NAME           0.00
NPD_FIELD_CODE               0.00
NPD_FIELD_NAME               0.00
NPD_FACILITY_CODE            0.00
NPD_FACILITY_NAME            0.00
ON_STREAM_HRS                1.82
AVG_DOWNHOLE_PRESSURE       42.56
AVG_DOWNHOLE_TEMPERATURE    42.56
AVG_DP_TUBING               42.56
AVG_ANNULUS_PRESS           49.53
AVG_CHOKE_SIZE_P            42.95
AVG_CHOKE_UOM               41.40
AVG_WHP_P                   41.44
AVG_WHT_P                   41.50
DP_CHOKE_SIZE                1.88
BORE_OIL_VOL                41.40
BORE_GAS_VOL                41.40
BORE_WAT_VOL                41.40
BORE_WI_VOL                 63.50
FLOW_KIND                    0.00
WELL_TYPE                    0.00
dtype: float64

In [9]:
data.isna().sum().sum() / data.size * 100

22.33033772547013

22.33% of our data is NaN values. <br />
Percentage NaN values by columns is given below
| Column                      | Percentage |
|-----------------------------|------------|
| DATEPRD                     | 0.00       |
| WELL_BORE_CODE              | 0.00       |
| NPD_WELL_BORE_CODE          | 0.00       |
| NPD_WELL_BORE_NAME          | 0.00       |
| NPD_FIELD_CODE              | 0.00       |
| NPD_FIELD_NAME              | 0.00       |
| NPD_FACILITY_CODE           | 0.00       |
| NPD_FACILITY_NAME           | 0.00       |
| ON_STREAM_HRS               | 1.82       |
| AVG_DOWNHOLE_PRESSURE       | 42.56      |
| AVG_DOWNHOLE_TEMPERATURE    | 42.56      |
| AVG_DP_TUBING               | 42.56      |
| AVG_ANNULUS_PRESS           | 49.53      |
| AVG_CHOKE_SIZE_P            | 42.95      |
| AVG_CHOKE_UOM               | 41.40      |
| AVG_WHP_P                   | 41.44      |
| AVG_WHT_P                   | 41.50      |
| DP_CHOKE_SIZE               | 1.88       |
| BORE_OIL_VOL                | 41.40      |
| BORE_GAS_VOL                | 41.40      |
| BORE_WAT_VOL                | 41.40      |
| BORE_WI_VOL                 | 63.50      |
| FLOW_KIND                   | 0.00       |
| WELL_TYPE                   | 0.00       |


In [10]:
np.unique(data['NPD_WELL_BORE_CODE'])

array([5351, 5599, 5693, 5769, 7078, 7289, 7405], dtype=int64)

In [11]:
grouped_data = data.groupby('NPD_WELL_BORE_CODE')

In [12]:
dataframes = {well_bore_code : group for well_bore_code, group in grouped_data}

In [13]:
data_well1, data_well2, data_well3, data_well4, data_well5, data_well6, data_well7 = dataframes.values()

In [14]:
print(data_well1.shape, data_well2.shape, data_well3.shape, data_well4.shape, data_well5.shape, data_well6.shape, data_well7.shape)

(3056, 24) (3056, 24) (3327, 24) (3306, 24) (1165, 24) (978, 24) (746, 24)


## Analysis for well NO 15/9-F-14 H || 5351 starts below

In [15]:
data_well1.head()

DATEPRD  WELL_BORE_CODE  NPD_WELL_BORE_CODE NPD_WELL_BORE_NAME  \
4967 2008-02-12  NO 15/9-F-14 H                5351          15/9-F-14   
4968 2008-02-13  NO 15/9-F-14 H                5351          15/9-F-14   
4969 2008-02-14  NO 15/9-F-14 H                5351          15/9-F-14   
4970 2008-02-15  NO 15/9-F-14 H                5351          15/9-F-14   
4971 2008-02-16  NO 15/9-F-14 H                5351          15/9-F-14   

      NPD_FIELD_CODE NPD_FIELD_NAME  NPD_FACILITY_CODE NPD_FACILITY_NAME  \
4967         3420717          VOLVE             369304    MÆRSK INSPIRER   
4968         3420717          VOLVE             369304    MÆRSK INSPIRER   
4969         3420717          VOLVE             369304    MÆRSK INSPIRER   
4970         3420717          VOLVE             369304    MÆRSK INSPIRER   
4971         3420717          VOLVE             369304    MÆRSK INSPIRER   

      ON_STREAM_HRS  AVG_DOWNHOLE_PRESSURE  AVG_DOWNHOLE_TEMPERATURE  \
4967            0.0                    0.0                       0.0   
4968            0.0                    0.0                       0.0   
4969            0.0                    0.0                       0.0   
4970            0.0                    0.0                       0.0   
4971            0.0                    0.0                       0.0   

      AVG_DP_TUBING  AVG_ANNULUS_PRESS  AVG_CHOKE_SIZE_P AVG_CHOKE_UOM  \
4967            0.0            0.00000               NaN             %   
4968            0.0            0.00000               NaN             %   
4969            0.0            0.00000               NaN             %   
4970            0.0            0.00006               NaN             %   
4971            0.0            0.00006               NaN             %   

      AVG_WHP_P  AVG_WHT_P  DP_CHOKE_SIZE  BORE_OIL_VOL  BORE_GAS_VOL  \
4967        0.0        0.0        0.05885           0.0           0.0   
4968        0.0        0.0        0.06768           0.0           0.0   
4969        0.0        0.0        0.04950           0.0           0.0   
4970        0.0        0.0        0.06640           0.0           0.0   
4971        0.0        0.0        0.10479           0.0           0.0   

      BORE_WAT_VOL  BORE_WI_VOL   FLOW_KIND WELL_TYPE  
4967           0.0          NaN  production        OP  
4968           0.0          NaN  production        OP  
4969           0.0          NaN  production        OP  
4970           0.0          NaN  production        OP  
4971           0.0          NaN  production        OP

In [16]:
data_well1.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3056 entries, 4967 to 8022
Data columns (total 24 columns):
 #   Column                    Non-Null Count  Dtype         
---  ------                    --------------  -----         
 0   DATEPRD                   3056 non-null   datetime64[ns]
 1   WELL_BORE_CODE            3056 non-null   object        
 2   NPD_WELL_BORE_CODE        3056 non-null   int64         
 3   NPD_WELL_BORE_NAME        3056 non-null   object        
 4   NPD_FIELD_CODE            3056 non-null   int64         
 5   NPD_FIELD_NAME            3056 non-null   object        
 6   NPD_FACILITY_CODE         3056 non-null   int64         
 7   NPD_FACILITY_NAME         3056 non-null   object        
 8   ON_STREAM_HRS             3056 non-null   float64       
 9   AVG_DOWNHOLE_PRESSURE     3050 non-null   float64       
 10  AVG_DOWNHOLE_TEMPERATURE  3050 non-null   float64       
 11  AVG_DP_TUBING             3050 non-null   float64       
 12  AVG_ANNULUS_PRESS     

In [17]:
data_well1.nunique()

DATEPRD                     3056
WELL_BORE_CODE                 1
NPD_WELL_BORE_CODE             1
NPD_WELL_BORE_NAME             1
NPD_FIELD_CODE                 1
NPD_FIELD_NAME                 1
NPD_FACILITY_CODE              1
NPD_FACILITY_NAME              1
ON_STREAM_HRS                280
AVG_DOWNHOLE_PRESSURE       2837
AVG_DOWNHOLE_TEMPERATURE    2733
AVG_DP_TUBING               2900
AVG_ANNULUS_PRESS           1482
AVG_CHOKE_SIZE_P            1672
AVG_CHOKE_UOM                  1
AVG_WHP_P                   2894
AVG_WHT_P                   2862
DP_CHOKE_SIZE               3041
BORE_OIL_VOL                2705
BORE_GAS_VOL                2723
BORE_WAT_VOL                2676
BORE_WI_VOL                    0
FLOW_KIND                      1
WELL_TYPE                      1
dtype: int64

In [18]:
data_well1.isna().sum()

DATEPRD                        0
WELL_BORE_CODE                 0
NPD_WELL_BORE_CODE             0
NPD_WELL_BORE_NAME             0
NPD_FIELD_CODE                 0
NPD_FIELD_NAME                 0
NPD_FACILITY_CODE              0
NPD_FACILITY_NAME              0
ON_STREAM_HRS                  0
AVG_DOWNHOLE_PRESSURE          6
AVG_DOWNHOLE_TEMPERATURE       6
AVG_DP_TUBING                  6
AVG_ANNULUS_PRESS            523
AVG_CHOKE_SIZE_P             196
AVG_CHOKE_UOM                  0
AVG_WHP_P                      0
AVG_WHT_P                      0
DP_CHOKE_SIZE                  0
BORE_OIL_VOL                   0
BORE_GAS_VOL                   0
BORE_WAT_VOL                   0
BORE_WI_VOL                 3056
FLOW_KIND                      0
WELL_TYPE                      0
dtype: int64

In [19]:
data_well1.isna().mean().round(4) * 100

DATEPRD                       0.00
WELL_BORE_CODE                0.00
NPD_WELL_BORE_CODE            0.00
NPD_WELL_BORE_NAME            0.00
NPD_FIELD_CODE                0.00
NPD_FIELD_NAME                0.00
NPD_FACILITY_CODE             0.00
NPD_FACILITY_NAME             0.00
ON_STREAM_HRS                 0.00
AVG_DOWNHOLE_PRESSURE         0.20
AVG_DOWNHOLE_TEMPERATURE      0.20
AVG_DP_TUBING                 0.20
AVG_ANNULUS_PRESS            17.11
AVG_CHOKE_SIZE_P              6.41
AVG_CHOKE_UOM                 0.00
AVG_WHP_P                     0.00
AVG_WHT_P                     0.00
DP_CHOKE_SIZE                 0.00
BORE_OIL_VOL                  0.00
BORE_GAS_VOL                  0.00
BORE_WAT_VOL                  0.00
BORE_WI_VOL                 100.00
FLOW_KIND                     0.00
WELL_TYPE                     0.00
dtype: float64

In [20]:
round(data_well1.describe(include=['int64', 'float64']).T, 2)

count        mean        std         min  \
NPD_WELL_BORE_CODE        3056.0     5351.00       0.00     5351.00   
NPD_FIELD_CODE            3056.0  3420717.00       0.00  3420717.00   
NPD_FACILITY_CODE         3056.0   369304.00       0.00   369304.00   
ON_STREAM_HRS             3056.0       20.54       7.88        0.00   
AVG_DOWNHOLE_PRESSURE     3050.0      233.07      64.92        0.00   
AVG_DOWNHOLE_TEMPERATURE  3050.0       95.13      25.85        0.00   
AVG_DP_TUBING             3050.0      192.65      57.74        0.00   
AVG_ANNULUS_PRESS         2533.0       10.53       9.70        0.00   
AVG_CHOKE_SIZE_P          2860.0       69.39      31.06        0.00   
AVG_WHP_P                 3056.0       41.53      22.72        0.00   
AVG_WHT_P                 3056.0       77.10      25.62        0.00   
DP_CHOKE_SIZE             3056.0       15.44      19.97        0.00   
BORE_OIL_VOL              3056.0     1290.00    1298.36        0.00   
BORE_GAS_VOL              3056.0   189139.25  184204.11        0.00   
BORE_WAT_VOL              3056.0     2330.25    1462.93      -59.19   
BORE_WI_VOL                  0.0         NaN        NaN         NaN   

                                 25%         50%         75%         max  
NPD_WELL_BORE_CODE           5351.00     5351.00     5351.00     5351.00  
NPD_FIELD_CODE            3420717.00  3420717.00  3420717.00  3420717.00  
NPD_FACILITY_CODE          369304.00   369304.00   369304.00   369304.00  
ON_STREAM_HRS                  24.00       24.00       24.00       25.00  
AVG_DOWNHOLE_PRESSURE         237.49      246.78      261.97      334.66  
AVG_DOWNHOLE_TEMPERATURE       99.62      101.01      105.05      106.77  
AVG_DP_TUBING                 180.75      204.25      229.63      302.11  
AVG_ANNULUS_PRESS               0.00       12.97       21.23       26.68  
AVG_CHOKE_SIZE_P               43.01       71.01      100.00      100.00  
AVG_WHP_P                      31.02       33.55       49.08      125.72  
AVG_WHT_P                      81.22       86.96       88.54       93.51  
DP_CHOKE_SIZE                   2.67        7.18       21.17      125.72  
BORE_OIL_VOL                  209.86      880.78     2033.70     5644.37  
BORE_GAS_VOL                31304.04   142362.86   305327.86   789974.73  
BORE_WAT_VOL                  695.59     2965.72     3444.15     5691.77  
BORE_WI_VOL                      NaN         NaN         NaN         NaN

In [21]:
color_dict = {
    'BORE_OIL_VOL': 'forestgreen',
    'BORE_GAS_VOL': 'red',
    'BORE_WAT_VOL': 'royalblue'
}

In [22]:
y = data_well1.select_dtypes(exclude='object').columns
y
fig = px.line(data_well1, 
              x = 'DATEPRD', 
              y = y[4:], 
              log_y=True, 
              title = 'Trends of well NO 15/9-F-14 H',
              height = 800,
              color_discrete_map=color_dict
              )
fig.update_layout(xaxis_title = 'DATE', yaxis_title = 'Log Observation')
fig.update_xaxes(rangeslider_visible = True)

fig.show()

In [23]:
fig = px.imshow(data_well1[['BORE_OIL_VOL', 'BORE_GAS_VOL', 'BORE_WAT_VOL']].corr(method='spearman'),
                text_auto=True,
                color_continuous_scale=px.colors.sequential.Greens,
                x=['Oil production', 'Gas Production', 'Water Production'],
                y=['Oil production', 'Gas Production', 'Water Production'],
                height = 800,
                width = 800,
                title = 'Correlation between Oil, Gas and Water production'
)
fig.update_xaxes(side = 'top')
fig.show()

In [24]:
fig = px.imshow(data_well1[['BORE_OIL_VOL', 'BORE_GAS_VOL', 'BORE_WAT_VOL', 'AVG_CHOKE_SIZE_P', 'DP_CHOKE_SIZE', 'AVG_DP_TUBING', 'AVG_DOWNHOLE_PRESSURE', 'AVG_WHP_P']].corr(method='spearman'),
                text_auto=True,
                color_continuous_scale=px.colors.sequential.Greens,
                # x=['Oil production', 'Gas Production', 'Water Production'],
                # y=['Oil production', 'Gas Production', 'Water Production'],
                height = 900,
                width = 900,
                title = 'Correlation between Oil, Gas and Water production and other features'
)
fig.update_xaxes(side = 'top')
fig.show()

In [25]:
# sum_pr_ch = []
# sum_ch = []
# num = 0
# den = 0

In [26]:
# for y in data_temp['YEAR']:
#     for m in data_temp['MONTH']:
#         for d in data_temp['DAY']:
#             num = num + data_temp['PR_CH']
#             den = den + data_temp['AVG_CHOKE_SIZE_P']
#         sum_pr_ch.append(num)
#         sum_ch.append(den)
#         num = 0
#         den = 0

In [27]:
# flag = 0
# curr_mon = 0
# for y in data_temp['DATEPRD']:
#     if curr_mon == y.month:
#         flag = 0
#         print(y.day, y.month, y.year, flag)
#     else:
#         flag = 1
#         print(y.day, y.month, y.year, flag)
#         curr_mon = y.month


In [28]:
# sum_pr_ch = []
# sum_ch = []
# num = 0
# den = 0

In [29]:
# curr_mon = 0
# for y in data_temp['DATEPRD']:
#     if curr_mon == y.month:
#         num = num + data_temp['PR_CH']
#         den = den + data_temp['AVG_CHOKE_SIZE_P']
#     else:
#         sum_pr_ch.append(num)
#         sum_ch.append(den)
#         num = 0
#         den = 0
#         curr_mon = y.month


In [103]:
def process(data_in): # General helper function to preprocess data of each well

    data_copy = data_in.copy()

    for col in data_copy.columns:
        if data_copy[col].nunique() == 0 or data_copy[col].nunique() == 1:
            data_copy.drop(columns = col, inplace = True)

    y = data_copy[['BORE_GAS_VOL', 'BORE_OIL_VOL', 'BORE_WAT_VOL']]
    X = data_copy.drop(columns = y.columns)

    imp = KNNImputer(add_indicator=True, 
                     weights='distance', 
                     n_neighbors=10).set_output(transform='pandas')
    
    num_cols = make_column_selector(dtype_include=['int64', 'float64'])
    cat_cols = make_column_selector(dtype_include=['object', 'bool'])
    # date_col = make_column_selector(dtype_include=['datetime64[ns]'])
    
    ohe = OneHotEncoder(drop='first',
                        handle_unknown='ignore').set_output(transform='pandas')

    sc = StandardScaler()

    data_copy_x_train, data_copy_x_test, data_copy_y_train, data_copy_y_test = train_test_split(X, y,
                                                                                                train_size=0.8, 
                                                                                                test_size=0.2, 
                                                                                                random_state=seed)

    imp_x_train = imp.fit_transform(data_copy_x_train[num_cols])
    imp_x_test  = imp.transform(data_copy_x_test[num_cols])
    imp_y_train = imp.fit_transform(data_copy_y_train[num_cols])
    imp_y_test  = imp.transform(data_copy_y_test[num_cols])

    imp_x_train['DATEPRD'] = data_copy_x_train['DATEPRD']
    imp_x_test['DATEPRD'] = data_copy_x_test['DATEPRD']
    
    col_tx = make_column_transformer(
        (sc, num_cols),
        (ohe, cat_cols),
        verbose_feature_names_out=False,
        sparse_threshold=0,
        remainder='passthrough').set_output(transform='pandas')

    x_train_scaled = col_tx.fit_transform(imp_x_train)
    x_test_scaled  = col_tx.transform(imp_x_test)
    # y_train_scaled = col_tx.fit_transform(imp_y_train)
    # y_test_scaled  = col_tx.transform(imp_y_test)

    day_sin_x_train = np.sin(2 * np.pi * imp_x_train['DATEPRD'].dt.day_of_year / 365)
    day_cos_x_train = np.cos(2 * np.pi * imp_x_train['DATEPRD'].dt.day_of_year / 365)
    
    month_sin_x_train = np.sin(2 * np.pi * imp_x_train['DATEPRD'].dt.month / 12)
    month_cos_x_train = np.cos(2 * np.pi * imp_x_train['DATEPRD'].dt.month / 12)
    
    day_sin_x_test = np.sin(2 * np.pi * imp_x_test['DATEPRD'].dt.day_of_year / 365)
    day_cos_x_test = np.cos(2 * np.pi * imp_x_test['DATEPRD'].dt.day_of_year / 365)
    
    month_sin_x_test = np.sin(2 * np.pi * imp_x_test['DATEPRD'].dt.month / 12)
    month_cos_x_test = np.cos(2 * np.pi * imp_x_test['DATEPRD'].dt.month / 12)

    x_train_scaled['day_sin'] = day_sin_x_train
    x_train_scaled['day_cos'] = day_cos_x_train
    x_train_scaled['month_sin'] = month_sin_x_train
    x_train_scaled['month_cos'] = month_cos_x_train
    
    x_test_scaled['day_sin'] = day_sin_x_test
    x_test_scaled['day_cos'] = day_cos_x_test
    x_test_scaled['month_sin'] = month_sin_x_test
    x_test_scaled['month_cos'] = month_cos_x_test

    x_train_scaled.drop(columns = ['DATEPRD'], inplace=True)
    x_test_scaled.drop(columns = ['DATEPRD'], inplace=True)

    
    # return x_train_scaled, x_test_scaled, y_train_scaled, y_test_scaled
    return x_train_scaled, x_test_scaled, imp_y_train, imp_y_test

In [104]:
well1_x_train, well1_x_test, well1_y_train, well1_y_test = process(data_well1)

In [105]:
well1_x_train

ON_STREAM_HRS  AVG_DOWNHOLE_PRESSURE  AVG_DOWNHOLE_TEMPERATURE  \
6391       0.438535              -0.198129                  0.371251   
5108      -2.591220               0.487575                 -0.016159   
6517       0.438535              -0.037772                  0.280239   
7278       0.438535               0.430231                  0.192805   
7376       0.438535               0.443145                  0.182059   
...             ...                    ...                       ...   
5730       0.438535               0.067472                  0.438106   
5802       0.438535               0.232948                  0.454556   
6620       0.438535               0.023525                  0.261945   
7574       0.438535               0.540461                  0.181588   
7699       0.438535               0.548084                  0.175875   

      AVG_DP_TUBING  AVG_ANNULUS_PRESS  AVG_CHOKE_SIZE_P  AVG_WHP_P  \
6391      -0.190757          -0.925910         -0.587312  -0.137010   
5108       1.215349          -0.399313         -1.866235  -1.754914   
6517       0.086245          -0.925910          1.001388  -0.382453   
7278       0.634477           1.331007          0.886463  -0.431458   
7376       0.643293           0.245994          1.001388  -0.416623   
...             ...                ...               ...        ...   
5730      -0.139985          -0.925910         -0.516445   0.502908   
5802       0.145680          -0.925910          0.160507   0.250094   
6620       0.081473           1.390018         -0.199314  -0.192478   
7574       0.766794           0.642913          1.001388  -0.451184   
7699       0.768569           0.370208          0.309698  -0.433634   

      AVG_WHT_P  DP_CHOKE_SIZE  missingindicator_AVG_DOWNHOLE_PRESSURE  \
6391   0.324491      -0.415728                               -0.040489   
5108  -2.413487      -0.016814                               -0.040489   
6517   0.338280      -0.602622                               -0.040489   
7278   0.442379      -0.652213                               -0.040489   
7376   0.352329      -0.667374                               -0.040489   
...         ...            ...                                     ...   
5730   0.422641       0.203359                               -0.040489   
5802   0.615137      -0.091038                               -0.040489   
6620   0.427141      -0.407112                               -0.040489   
7574   0.460489      -0.652775                               -0.040489   
7699   0.225084      -0.585662                               -0.040489   

      missingindicator_AVG_DOWNHOLE_TEMPERATURE  \
6391                                  -0.040489   
5108                                  -0.040489   
6517                                  -0.040489   
7278                                  -0.040489   
7376                                  -0.040489   
...                                         ...   
5730                                  -0.040489   
5802                                  -0.040489   
6620                                  -0.040489   
7574                                  -0.040489   
7699                                  -0.040489   

      missingindicator_AVG_DP_TUBING  missingindicator_AVG_ANNULUS_PRESS  \
6391                       -0.040489                           -0.441713   
5108                       -0.040489                           -0.441713   
6517                       -0.040489                           -0.441713   
7278                       -0.040489                           -0.441713   
7376                       -0.040489                           -0.441713   
...                              ...                                 ...   
5730                       -0.040489                           -0.441713   
5802                       -0.040489                           -0.441713   
6620                       -0.040489                           -0.441713   
7574                      

In [108]:
fig = px.scatter(x = well1_x_train['day_sin'], y = well1_x_train['day_cos'], width = 500, title='Cyclic nature of days of year')
fig.show()
fig = px.scatter(x = well1_x_train['month_sin'], y = well1_x_train['month_cos'], width = 500, title='Cyclic nature of months')
fig.show()